### DISCUSSION TITLE
Right-Align Text in Specific DataGrid Column Headers

### DISCUSSION THREAD

**Discussion by MichaelSvejcar** [AvaloniaUI/Avalonia#11769] [Q&A]
I'm currently working on a project where I need to right-align the text in certain DataGrid column headers. However, I'm having trouble achieving this.

I've tried using a HeaderTemplate with a TextBlock and setting the TextAlignment to "Right", but this results in the header text not being displayed at all.

I also attempted to set a style in the DataGrid.Resources for DataGridColumnHeader with a Setter for HorizontalContentAlignment set to "Right". However, this seems to apply the alignment to all column headers, not just the specific ones I'm targeting.

I've looked into using x:Name to target specific column headers, but I get an error stating "Unable to resolve suitable regular or attached property Name on type Avalonia.Controls.DataGrid:Avalonia.Controls.DataGridTemplateColumn".

I'm not sure how to proceed. Could anyone guide me on how to right-align the text in specific DataGrid column headers? Any help or pointers would be greatly appreciated. Thanks in advance.

**Comment by timunie** [comment]
1. Which Avalonia version? 
2. Which Theme do you use?

In general: It's often a good idea to look up original styles. I guess you'll have to redefine `ControlTemplate` to get it working.

**Reply by MichaelSvejcar** [reply]
1) Avalonia 0.10.21
2) Fluent theme.

Is there really no other way than redefining ControlTemplate? If this is what I need to do, how should I proceed?

**Reply by timunie** [reply]
look up original style in source: https://github.com/AvaloniaUI/Avalonia/blob/d59d66e07de6183b5cbbb2bec5b4d1d427331ee5/src/Avalonia.Controls.DataGrid/Themes/Fluent.xaml#L186C33-L239 

for 0.10.x, you have to switch branch as master is for 11.0 only.

**Reply by MichaelSvejcar** [reply]
Thank you for the response. To be honest, I am still quite confused, as the HorizontalContentAlignment is defined under the <Style Selector="DataGridColumnHeader">, not under the ControlTemplate.
Does that mean, that I need to somehow create new Style for DataGridColumnHeader, in which the HorizontalContentAlignment will be set to right, and then use this one specific style only for the columns, in which I want this to be applied? If not, how should I implement this?
Thank you very much.

**Reply by timunie** [reply]
the solution is actually easy: 

```xml
<DataGrid Items="{Binding People}" CanUserResizeColumns="True">
    <DataGrid.Columns>
        <DataGridTextColumn Binding="{Binding (vm:Person).Name}"></DataGridTextColumn>
        <DataGridTextColumn Binding="{Binding (vm:Person).Age}" Width="200">
            <DataGridTextColumn.HeaderTemplate>
                <!-- Simply use a HeaderTemplate --> 
                <DataTemplate>
                    <TextBlock Text="Right Text" HorizontalAlignment="Right"></TextBlock>
                </DataTemplate>
            </DataGridTextColumn.HeaderTemplate>
        </DataGridTextColumn>
    </DataGrid.Columns>
</DataGrid>
```

if you want to get rid of the sort icon reserved space, you still have to modify the control template: 

```diff
<Style Selector="DataGridColumnHeader">
    <Setter Property="Template">
        <ControlTemplate>
            <Grid Name="PART_ColumnHeaderRoot"
                  ColumnDefinitions="*,Auto"
                  Background="{TemplateBinding Background}">

                <Grid HorizontalAlignment="{TemplateBinding HorizontalContentAlignment}"
                      VerticalAlignment="{TemplateBinding VerticalContentAlignment}"
                      Margin="{TemplateBinding Padding}">
                    <Grid.ColumnDefinitions>
                        <ColumnDefinition Width="*" />
!                       <ColumnDefinition MinWidth="0"
                                          Width="Auto" />
                    </Grid.ColumnDefinitions>

                    <ContentPresenter Content="{TemplateBinding Content}"
                                      ContentTemplate="{TemplateBinding ContentTemplate}" />

                    <Path Name="SortIcon"
                          Grid.Column="1"
                          Fill="{TemplateBinding Foreground}"
                          HorizontalAlignment="Center"
                          VerticalAlignment="Center"
                          Stretch="Uniform"
                          Height="12" />
                </Grid>

                <Rectangle Name="VerticalSeparator"
                           Grid.Column="1"
                           Width="1"
                           VerticalAlignment="Stretch"
                           Fill="{TemplateBinding SeparatorBrush}"
                           IsVisible="{TemplateBinding AreSeparatorsVisible}" />

                <Grid x:Name="FocusVisual"
                      IsHitTestVisible="False">
                    <Rectangle x:Name="FocusVisualPrimary"
                               HorizontalAlignment="Stretch"
                               VerticalAlignment="Stretch"
                               Fill="Transparent"
                               IsHitTestVisible="False"
                               Stroke="{DynamicResource DataGridCellFocusVisualPrimaryBrush}"
                               StrokeThickness="2" />
                    <Rectangle x:Name="FocusVisualSecondary"
                               Margin="2"
                               HorizontalAlignment="Stretch"
                               VerticalAlignment="Stretch"
                               Fill="Transparent"
                               IsHitTestVisible="False"
                               Stroke="{DynamicResource DataGridCellFocusVisualSecondaryBrush}"
                               StrokeThickness="1" />
                </Grid>
            </Grid>
        </ControlTemplate>
    </Setter>
</Style>

```

**Reply by MichaelSvejcar** [reply]
Thank you very much for your detailed response!

However, I have implemented it exactly like this, and the outcome is that the header text is not visible at all.
In order to debug this, I also tried to set Background of the TextBlock to Yellow, to see if the TextBlock is visible at all, but still, nothing visible at all.
What can be causing this issue? I am using Avalonia.Controls.DataGrid v0.10.21.

Wish you a nice weekend!
